src/main_context_channel.rs

```diff
@@ -0,0 +1,439 @@
+
+        let sender = self.0.as_ref().expect("No Sender anymore");
+        sender.try_send(t)?;
+
+        // Once sending succeeded, wake up the receiver
+        unsafe {
+            let source = &*(self.1.to_glib_none().0 as *const ChannelSource<T>);
+            source.mark_ready();
+        }
+        Ok(())
+    }
+}
+
+impl<T> Drop for SyncSender<T> {
+    fn drop(&mut self) {
+        // Wake up the Source so that it can be removed if this was the last sender.
+        // We have to drop the Sender first because of that
+        drop(self.0.take().expect("No Sender anymore"));
+
+        unsafe {
+            let source = &*(self.1.to_glib_none().0 as *const ChannelSource<T>);
+            source.mark_ready();
+        }
+    }
+}
+
+/// A `Receiver` that can be attached to a main context to receive items from its corresponding
+/// `Sender` or `SyncSender`.
+///
+/// See [`MainContext::channel()`] or [`MainContext::sync_channel()`] for how to create
+/// such a `Receiver`.
+///
+/// [`MainContext::channel()`]: struct.MainContext.html#method.channel
+/// [`MainContext::sync_channel()`]: struct.MainContext.html#method.sync_channel
+#[derive(Debug)]
+pub struct Receiver<T>(Option<Source>, PhantomData<*const T>);
+
+unsafe impl<T: Send> Send for Receiver<T> {}
+
+impl<T> Drop for Receiver<T> {
+    fn drop(&mut self) {
+        // If the receiver was never attached to a main context
+        // we need to destroy the underlying source
+        if let Some(source) = self.0.take() {
+            source.destroy();
+        }
+    }
+}
+
+impl<T> Receiver<T> {
+    /// Attaches the receiver to the given `context` and calls `func` whenever an item is
+    /// available.
+    ///
+    /// Passing `None` for the context will attach it to the thread default main context.
+    ///
+    /// # Panics
+    ///
+    /// This function panics if called from a thread that is not the owner of the provided
+    /// `context`, or if `None` is provided of the thread default main context.
+    pub fn attach<'a, P: Into<Option<&'a MainContext>>, F: FnMut(T) -> Continue + 'static>(
+        mut self,
+        context: P,
+        func: F,
+    ) -> u32 {
+        let context = context.into();
+        unsafe {
+            let source = self.0.take().expect("No Source anymore");
+
+            {
+                let source = &mut *(source.to_glib_none().0 as *mut ChannelSource<T>);
+                source.callback = Some(RefCell::new(Box::new(func)));
+                source.thread_id = Some(get_thread_id());
+            }
+
+            let id = if let Some(context) = context {
+                assert!(context.is_owner());
+                source.attach(context)
+            } else {
+                let context = MainContext::ref_thread_default();
+                assert!(context.is_owner());
+                source.attach(&context)
+            };
+
+            id
+        }
+    }
+}
+
+impl MainContext {
+    /// Create a channel for this main context
```

Comment:
Missing dot at the end.

Also, you usually use the singular third pronoun everywhere else but not here. I was expecting "Creates". :)